hamilflow/models/pendulum.py

```diff
@@ -0,0 +1,128 @@
+import math
+from functools import cached_property
+
+import numpy as np
+import pandas as pd
+from numpy.typing import ArrayLike
+from pydantic import BaseModel, Field, computed_field
+from scipy.special import ellipj, ellipk
+
+
+class PendulumSystem(BaseModel):
+    r"""The params for the pendulum
+
+    :param omega0 (float): $\ometa_0$, frequency parameter in the action
+    """
+
+    omega0: float = Field(default=2 * math.pi, gt=0)
+
+
+class PendulumIC(BaseModel):
+    r"""The initial condition for a pendulum
+
+    :param theta0 (float): $\theta_0$, the initial angular displacement
+    """
+
+    theta0: float = Field(ge=-math.pi / 2, le=math.pi / 2)
+
+    @computed_field  # type: ignore[misc]
+    @cached_property
+    def k(self) -> float:
+        r"""$k\coloneqq\sin(\theta_0 / 2)$"""
+        return math.sin(self.theta0 / 2)
+
+
+class Pendulum(BaseModel):
+    r"""Generate time series data for a pendulum.
+
+    # Lagrangian action
+    The Lagrangian action for a pendulum is
+    $$
+    S\[\theta(t)\] = \int_{0}^{t_0} \mathbb{d}t
+    \left\\{\frac{1}{2} I \dot\theta^2 + U \cos\theta \right\\} \eqqcolon
+    \int_{0}^{t_0} \mathbb{d}t\,L_\text{P}(\theta, \dot\theta)\,,
+    $$
+    where $\theta$ is the angular displacement from the vertical to the
+    pendulum; $I$ is an _inertial parameter_, $U$ is a potential parameter;
+    $L_\text{P}$ is the Lagrangian.
+
+    This setup contains both the single and the physical pendula. For a single
+    pendulum,
+    $$
+    I = m l^2\,,\qquad U = mgl\,,
+    $$
+    where $m$ is the mass of the pendulum, $l$ is the length of the rod or cord,
+    and $g$ is the gravitational acceleration.
+
+    # Integral of motion
+    $\mathbb{\delta}S / \mathbb{\delta}{t} = 0$
+    $$
+    \dot\theta\frac{\partial L_\text{P}}{\partial \dot\theta} - L_\text{P}
+    \equiv E = U \cos\theta_0
+    $$
+
+    $$
+    \left(\frac{\mathbb{d}t}{\mathbb{d}\theta}\right)^2 = \frac{1}{2\omega_0^2}
+    \frac{1}{\cos\theta - \cos\theta_0}
+    $$
+    $\omega_0 \coloneqq \sqrt{\frac{U}{I}}$
+
+    ## Coordinate transformation
+    $$
+    \sin u \coloneqq \frac{\sin\frac{\theta}{2}}{\sin\frac{\theta_0}{2}}
+    $$
+
+    $$
+    \left(\frac{\mathbb{d}t}{\mathbb{d}u}\right)^2 = \frac{1}{\omega_0^2}
+    \frac{1}{1-k^2\sin^2 u}
+    $$
+    """
+
+    system: PendulumSystem  # | float | Dict[str, float]
+    initial_condition: PendulumIC  # | float | Dict[str, float]
```

Comment:
@emptymalei how can I improve the initialiser interface if I wanna keep `Pendulum` as a subclass of `BaseModel`? At the moment I need to
```py
ps = PendulumSystem(omega0=omega0)
ic = PendulumIC(theta0=theta0)
p = Pendulum(system=ps, initial_condition=ic)
```
which is cubersome.

Reply:
I think there two things to be discussed here.

1. Shall `Pendulum` inherit from `BaseModel`?
2. Shall we instantiate the model `Pendulum` using BaseModel/dataclass?
    1. I see you are using classvar for system and IC. This will require us to create the System and IC as pydantic model beforehand. This also requires a customized serializer. I included some detailed discussions on the drawbacks.


## 1. Shall `Pendulum` inherit from `BaseModel`?

I prefer not to do this. 

BaseModel is like dataclass. We use this to keep a model like Pendulum modular. Meanwhile it is a good validation tool for the params.

Though it behaves just like a normal class apart from the dataclass capabilities, it doesn't benefit us on our actual model calculations.

Pydantic also does some stuff that we don't really need for `Pendulum`. I would prefer to be simple.


## 2. Shall we instantiate the model `Pendulum` using BaseModel/dataclass?

I am not sure which is the best. I had some thoughts when developed the first model.

### 2a. System Definition and Initial Condition passed to the Physics Model should be easy and fast

I would keep the init params of a model to be something native to Python (and I chose dictionary), this is why the initial conditions and system definitions are not exposed to the users 

```python
Pendulum(system={"a": ..., "b": ...}, initial_condition={...})
```

Comparing to 

```python
ps = PendulumSystem(omega0=omega0)
ic = PendulumIC(theta0=theta0)
p = Pendulum(system=ps, initial_condition=ic)
```

I think my version requires less cognitive load. But we can have a look at a few use cases.

I also see the benefit of using BaseModel to instantiate `Pendulum`, e.g., knowing the error early when feeding in the params. However, the benefits are marginal. Even if we use Dict as inputs, whenever an error occurs, it is clear to the user that a specific variable is problematic. 

### 2.b Use Case 1: New user walks in and creates a first model.

It will be easier for the user to use dict as they don't need to learn an additional concept called system definition dataclass/Pydantic and initial condition dataclass/Pydantic.

### 2.c Use Case 2: Generating Data in Batch

This is likely to be the most frequent use case.

When we generate datasets in batch, I am thinking of using json/yaml as configs.

```
configs = json.load(...) # code to read in the configs

for config in configs:
    MyModel(**config).save(...)
```

We just need to load the configs and there is no need to create `XXXSystem` and `XXXIC`.